firestore/api/SolutionCounter/Program.cs

```diff
@@ -0,0 +1,209 @@
+    {
+        /// <summary>
+        /// Gets or sets the Count
+        /// </summary>
+        [FirestoreProperty]
+        public int Count { get; set; }
+    }
+
+    // [END fs_counter_classes]
+
+    /// <summary>
+    /// Defines the <see cref="DistributedCounter" />
+    /// </summary>
+    public class DistributedCounter
+    { 
+        public static string Usage = @"Usage:
+C:\> dotnet run command YOUR_PROJECT_ID
+
+Where command is one of
+    run-distributed-counter
+";
+        /// <summary>
+        /// Gets or sets the FireStoreDb
+        /// </summary>
+        public static FirestoreDb FirestoreDb { get; set; }
+
+        /// <summary>
+        /// Gets or sets the ShardsCounter
+        /// </summary>
+        public static Counter ShardsCounter { get; set; }
+
+        /// <summary>
+        /// The InitFireStoreDb
+        /// </summary>
+        public static void InitFirestoreDb(string project)
+        {
+            var projectId = project;
+            if (string.IsNullOrEmpty(projectId))
+                throw new ArgumentException("ArgumentException: failed read Firestore Project ID");
+
+            FirestoreDb = FirestoreDb.Create(projectId: projectId);
+            Console.WriteLine($"Created Cloud Firestore client with project ID: {project}");
+        }
+
+        /// <summary>
+        /// The InitShardsCounter
+        /// </summary>
+        /// <param name="numShards">The numShards<see cref="int"/></param>
+        public static void InitShardsCounter(int numShards)
+        {
+            // Initialize whole global count of shards
+            // in ShardsCounter property 
+            ShardsCounter = new Counter
+            {
+                NumShards = numShards
+            };
+        }
+
+        // [START fs_create_counter]
+
+        /// <summary>
+        /// The InitCounterAsync creates a given number of shards as
+        /// sub collection of specified document.
+        /// </summary>
+        /// <param name="docRef">The docRef<see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        public static async Task InitCounterAsync(DocumentReference docRef)
+        {
+            CollectionReference colRef = docRef.Collection("shards");
+
+            // Initialize each shard with count=0
+            for (var i = 0; i < ShardsCounter.NumShards; i++)
+            {
+                await colRef.Document(i.ToString()).SetAsync(new Shard { Count = 0 });
+            }
+        }
+
+        // [END fs_create_counter]
+
+        // [START fs_increment_counter]
+
+        /// <summary>
+        /// The IncrementCounterAsync increments a randomly picked shard.
+        /// </summary>
+        /// <param name="docRef">The docRef<see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        public static async Task IncrementCounterAsync(DocumentReference docRef)
+        {
+            var rand = new Random();
```

Comment:
Even for a sample this should probabl be static and used within a lock.
See https://stackoverflow.com/questions/36376888.